### ISSUE TITLE
Ability to interrupt macro execution

### ISSUE THREAD

**Issue by bubblobill** [RPTools/maptool#1464]
**Is your feature request related to a problem? Please describe.**
When a macro runs off into a nasty loop firing off dozens of input prompts or just sits there and grinds quietly away, My only option is to kill MapTool. 

**Describe the solution you'd like**
I would like the ability to force the macro to stop. Preferably with a key combination that is available to laptops (no break key).

**Describe alternatives you've considered**
Don't forget to save.

**Comment by Azhrei** [comment]
Lol, I had this happen to me recently.  I opened up the log file and waited to see the "autosave complete" message, then I killed MT.  (I had to wait awhile as my autosave was set to 15 minutes, but I'd done a lot of typing and didn't want to lose it.)

There is already a fix for preventing a single missing variable from causing popups when the stat sheet is displayed; we need the same thing for macros.

**Comment by Phergus** [comment]
Just FYI,
[setMaxLoopIteration()](http://www.lmwcs.com/rptools/wiki/setMaxLoopIterations)
And somewhat related:
[setMaxRecursionDepth()](http://www.lmwcs.com/rptools/wiki/setMaxRecursionDepth)

**Comment by Azhrei** [comment]
Question:  If 10,000 is both the maximum and minimum value for `setMaxLoopIteration()`, what's the point? 😉

**Comment by Phergus** [comment]
What indeed is the point? The point is frozen, the beast is late out of Paddington. The point is taken. If La Fontaine's elk would spurn Tom Jones the engine must be our head, the dining car our esophagus, the guard's van our left lung, the cattle truck our shins, the first-class compartment the piece of skin at the nape of the neck and the level crossing an electric elk called Simon. The clarity is devastating. But where is the ambiguity? It's over there in a box. Shunt is saying the 8:15 from Gillingham when in reality he means the 8:13 from Gillingham. The train is the same only the time is altered. Ecce homo, ergo elk. La Fontaine knew his sister and knew her bloody well. The point is taken, the beast is moulting, the fluff gets up your nose. The illusion is complete; it is reality, the reality is illusion and the ambiguity is the only truth. But is the truth, as Hitchcock observes, in the box? No there isn't room, the ambiguity has put on weight. The point is taken, the elk is dead, the beast stops at Swindon, Chabrol stops at nothing, I'm having treatment and La Fontaine can get knotted.

Which is to say, ask Wolph as he wrote the description.  The code only sets a minimum of 10,000.

**Comment by JamzTheMan** [comment]
I always thought it would be nice to have macros run in a killable thread with a timeout of 30-60 seconds or something.

Kind of like days of old when your browser would warn you that JavaScript was running long and if you wanted to kill it or keep going.

Nothing sucks more than writing code only to have MapTool lock up because of testing and you foobar the code.

**Comment by Phergus** [comment]
That would be nice.

**Comment by bubblobill** [comment]
> Just FYI,
> [setMaxLoopIteration()](http://www.lmwcs.com/rptools/wiki/setMaxLoopIterations)
> And somewhat related:
> [setMaxRecursionDepth()](http://www.lmwcs.com/rptools/wiki/setMaxRecursionDepth)

I have a macro that seems to exponentially increase in time required at every iteration. about 4secs for the first, 16secs for 2, 50secs for 3, and overnight for 20. Setting the loop interations to something like 10 would stop it running forever but cripple every other loop.

**Comment by Phergus** [comment]
Is it doing anything with macros on tokens?   Do you have the Selection window visible?

**Comment by bubblobill** [comment]
Yes and yes. It is removing all macros and adding a new set. 2 of them have images as labels.

**Comment by Phergus** [comment]
There is some problem with updating the selection window and adding/deleting macros that causes it to take a very long time.
